### PR TITLE
CRF for keras 2.0

### PR DESCRIPTION
**Pull request by linxihui**
An Implementation of linear chain conditional random field (CRF), modified for keras 2.0 from previous PR https://github.com/fchollet/keras/pull/4646.

### CODE REVIEW

**Comment by fchollet** [comment]
This is a very large change adding a brand new feature. Typically the process for such a change would be to first ship it in Keras contrib and have users try it out for a couple of months. After this trial period, if a sufficient large group of users want it, include it as part of core Keras.

This is designed to:
- Avoid features that are too niche. All code/APIs in core Keras have a maintenance cost, so we try to only include those that useful to many.
- Give branch new code some time to mature through real-world usage, both in terms of UX/APIs and in terms of implementation.

File: examples/conll2000_chunking_crf.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+word2idx = dict((w, i) for i, w in enumerate(vocab))
+
+
+def process_data(data, maxlen=None, onehot=False):
+    if maxlen is None:
+        maxlen = max(len(s) for s in data)
+    x = [[w[0].lower() for w in s] for s in data]
+    y = [[w[1] for w in s] for s in data]
+    x = pad_sequences([[word2idx.get(w[0].lower(), 1) for w in s] for s in data], maxlen=maxlen)
+    y = pad_sequences([[class_labels.index(w[1])for w in s] for s in data], maxlen=maxlen)
```
Comment (nzw0301) [inline examples/conll2000_chunking_crf.py]: space after `(w[1])`

File: keras/layers/crf.py [hunk 2]
```diff
@@ -0,0 +1,540 @@
+import keras
```
Comment (nzw0301) [inline keras/layers/crf.py]: unused statement

File: keras/layers/crf.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+                                                 constraint=self.boundary_constraint)
+            self.right_boundary = self.add_weight((self.units,),
+                                                  name='right_boundary',
+                                                  initializer=self.boundary_initializer,
+                                                  regularizer=self.boundary_regularizer,
+                                                  constraint=self.boundary_constraint)
+        self.built = True
+
+    def call(self, X, mask=None):
+        input_length = K.int_shape(X)[1]
```
Comment (nzw0301) [inline keras/layers/crf.py]: unused variable

**Comment by yoookoo** [comment]
How can I use this CRF layer in image segmentation, can I set units to be an array?

**Comment by ahundt** [comment]
https://github.com/farizrahman4u/keras-contrib also see CONTRIBUTING.md

**Comment by fchollet** [comment]
Closing since expecting it to go to `keras-contrib`. May get included in core Keras in the future.

**Pull request closed without merging**